File: src/providers/open_meteo.ts
/* eslint-disable @typescript-eslint/naming-convention */
/**
 * Open-Meteo provider — zero-config global fallback.
 *
 * Free, no API key. Endpoint: api.open-meteo.com/v1/forecast
 * Requires lat/lon; resolves from the merged city lookup tables.
 * Priority 11 — lowest in the chain, activates when every higher-priority
 * provider has declined or failed.
 */

import {
  CN_CITIES,
  DE_DWD_CITIES,
  METOFFICE_CITIES,
  US_NWS_CITIES,
  WMO_CODE_MAP,
} from "../data-loader.js";
import { makeWeatherData } from "../models.js";
import {
  type IWeatherProvider,
  type Location,
  LocationNotSupportedError,
  ProviderError,
  WeatherCondition,
  type WeatherData,
} from "../types.js";

const BASE_URL = "https://api.open-meteo.com/v1/forecast";

const CURRENT_PARAMS =
  "temperature_2m,relative_humidity_2m,weather_code,wind_speed_10m,apparent_temperature";
const DAILY_PARAMS =
  "weather_code,temperature_2m_max,temperature_2m_min,precipitation_probability_max,sunrise,sunset";
const HAS_TIMEZONE_SUFFIX = /(?:Z|[+-]\d{2}:?\d{2})$/;

export class OpenMeteoProvider implements IWeatherProvider {
  readonly name = "open-meteo";
  readonly priority = 11;
  readonly supportsForecast = true;
  readonly requiresApiKey = false;

  supportsLocation(location: Location): boolean {
    return this.resolveCoords(location) !== null;
  }

  async getCurrent(location: Location): Promise<WeatherData> {
    const coords = this.resolveCoords(location);
    if (!coords) {
      throw new LocationNotSupportedError(
        `Open-Meteo: cannot resolve coordinates for: ${location.raw}`,
      );
    }
    const [lat, lon] = coords;
    const url = this.buildUrl(lat, lon, {
      current: CURRENT_PARAMS,
      daily: DAILY_PARAMS,
      forecast_days: "1",
      timezone: "UTC",
    });
    const data = await this.fetchJson(url);
    return this.parseCurrent(location, data);
  }

  async getForecast(location: Location, days: number = 7): Promise<WeatherData[]> {
    const coords = this.resolveCoords(location);
    if (!coords) {
      throw new LocationNotSupportedError(
        `Open-Meteo: cannot resolve coordinates for: ${location.raw}`,
      );
    }
    const [lat, lon] = coords;
    const url = this.buildUrl(lat, lon, {
      daily: DAILY_PARAMS,
      forecast_days: String(Math.min(days, 16)),
      timezone: "UTC",
    });
    const data = await this.fetchJson(url);
    return this.parseForecast(location, data, days);
  }

  // ── Private helpers ───────────────────────────────────────────────

  private resolveCoords(location: Location): [number, number] | null {
    if (location.latitude != null && location.longitude != null) {
      return [location.latitude, location.longitude];
    }
    const n = location.normalized;
    return (
      CN_CITIES[n] ??
      US_NWS_CITIES[n] ??
      DE_DWD_CITIES[n] ??
      METOFFICE_CITIES[n] ??
      null
    );
  }

  private buildUrl(
    lat: number,
    lon: number,
    extra: Record<string, string>,
  ): string {
    const params = new URLSearchParams({
      latitude: lat.toFixed(4),
      longitude: lon.toFixed(4),
      ...extra,
    });
    return `${BASE_URL}?${params.toString()}`;
  }

  private async fetchJson(url: string): Promise<Record<string, any>> {
    let res: Response;
    try {
      res = await fetch(url);
    } catch (e) {
      throw new ProviderError(
        `Open-Meteo request failed: ${(e as Error).message}`,
      );
    }
    if (!res.ok) {
      throw new ProviderError(`Open-Meteo API error: HTTP ${res.status}`);
    }
    return (await res.json()) as Record<string, any>;
  }

  private parseCurrent(
    location: Location,
    data: Record<string, any>,
  ): WeatherData {
    const current = (data.current ?? {}) as Record<string, any>;
    const daily = (data.daily ?? {}) as Record<string, any>;

    const wmoCode = current.weather_code ?? 0;
    const condition = WMO_CODE_MAP[String(wmoCode)] ?? WeatherCondition.Unknown;

    const tempHigh =
      Array.isArray(daily.temperature_2m_max) &&
      typeof daily.temperature_2m_max[0] === "number"
        ? daily.temperature_2m_max[0]
        : undefined;
    const tempLow =
      Array.isArray(daily.temperature_2m_min) &&
      typeof daily.temperature_2m_min[0] === "number"
        ? daily.temperature_2m_min[0]
        : undefined;

    const windSpeedKmh =
      typeof current.wind_speed_10m === "number"
        ? current.wind_speed_10m
        : undefined;

    const precipChance =
      Array.isArray(daily.precipitation_probability_max) &&
      typeof daily.precipitation_probability_max[0] === "number"
        ? daily.precipitation_probability_max[0]
        : undefined;

    const sunrise =
      Array.isArray(daily.sunrise) && typeof daily.sunrise[0] === "string"
        ? daily.sunrise[0]
        : undefined;
    const sunset =
      Array.isArray(daily.sunset) && typeof daily.sunset[0] === "string"
        ? daily.sunset[0]
        : undefined;

    return makeWeatherData({
      location: titleCase(location.normalized),
      temperature:
        typeof current.temperature_2m === "number"
          ? current.temperature_2m
          : 0,
      condition,
      condition_raw: `wmo:${wmoCode}`,
      provider_name: this.name,
      observed_at: this.parseObservedAt(current.time),
      latitude: typeof data.latitude === "number" ? data.latitude : undefined,
      longitude:
        typeof data.longitude === "number" ? data.longitude : undefined,
      ...(typeof current.relative_humidity_2m === "number"
        ? { humidity: current.relative_humidity_2m }
        : {}),
      ...(typeof current.apparent_temperature === "number"
        ? { feels_like: current.apparent_temperature }
        : {}),
      ...(windSpeedKmh != null ? { wind_speed: windSpeedKmh } : {}),
      ...(tempHigh != null ? { temp_high: tempHigh } : {}),
      ...(tempLow != null ? { temp_low: tempLow } : {}),
      ...(precipChance != null ? { precipitation_chance: precipChance } : {}),
      ...(sunrise ? { sunrise } : {}),
      ...(sunset ? { sunset } : {}),
    });
  }

  private parseForecast(
    location: Location,
    data: Record<string, any>,
    days: number,
  ): WeatherData[] {
    const daily = (data.daily ?? {}) as Record<string, any>;
    const times: string[] = Array.isArray(daily.time) ? daily.time : [];
    const wmoCodes: number[] = Array.isArray(daily.weather_code)
      ? daily.weather_code
      : [];
    const maxTemps: number[] = Array.isArray(daily.temperature_2m_max)
      ? daily.temperature_2m_max
      : [];
    const minTemps: number[] = Array.isArray(daily.temperature_2m_min)
      ? daily.temperature_2m_min
      : [];
    const precipChances: number[] = Array.isArray(
      daily.precipitation_probability_max,
    )
      ? daily.precipitation_probability_max
      : [];
    const sunrises: string[] = Array.isArray(daily.sunrise)
      ? daily.sunrise
      : [];
    const sunsets: string[] = Array.isArray(daily.sunset) ? daily.sunset : [];

    const locationName = titleCase(location.normalized);

    return times.slice(0, days).map((dateStr, i) => {
      const wmoCode = wmoCodes[i] ?? 0;
      const condition =
        WMO_CODE_MAP[String(wmoCode)] ?? WeatherCondition.Unknown;

      return makeWeatherData({
        location: locationName,
        temperature: 0,
        condition,
        condition_raw: `wmo:${wmoCode}`,
        provider_name: this.name,
        forecast_date: new Date(`${dateStr}T00:00:00.000Z`),
        ...(typeof maxTemps[i] === "number" ? { temp_high: maxTemps[i] } : {}),
        ...(typeof minTemps[i] === "number" ? { temp_low: minTemps[i] } : {}),
        ...(typeof precipChances[i] === "number"
          ? { precipitation_chance: precipChances[i] }
          : {}),
        ...(typeof sunrises[i] === "string" ? { sunrise: sunrises[i] } : {}),
        ...(typeof sunsets[i] === "string" ? { sunset: sunsets[i] } : {}),
      });
    });
  }

  private parseObservedAt(rawTime: unknown): Date {
    if (typeof rawTime !== "string") return new Date();
    const iso = HAS_TIMEZONE_SUFFIX.test(rawTime) ? rawTime : `${rawTime}Z`;
    const parsed = new Date(iso);
    return Number.isNaN(parsed.getTime()) ? new Date() : parsed;
  }
}

// ── Helpers ────────────────────────────────────────────────────────

function titleCase(s: string): string {
  return s
    .split(/[\s_-]+/)
    .map((w) => (w ? w[0]!.toUpperCase() + w.slice(1) : w))
    .join(" ");
}

File: weather/providers/open_meteo.py
"""
Open-Meteo provider — zero-config global fallback.

Free, no API key. Endpoint: api.open-meteo.com/v1/forecast
Requires lat/lon; resolves from the merged city lookup tables.
Priority 11 — lowest in the chain.
"""

import asyncio
import json
import urllib.parse
import urllib.request
from datetime import date, datetime, timezone
from typing import Optional

from ..data.loader import load_json
from ..models import Location, WeatherCondition, WeatherData
from .base import LocationNotSupportedError, ProviderError, WeatherProvider

# Load city coordinate tables — shared with Bun runtime.
_CN_CITIES: dict[str, tuple[float, float]] = {
    k: tuple(v) for k, v in load_json("cities", "cn.json").items()
}
_US_NWS_CITIES: dict[str, tuple[float, float]] = {
    k: tuple(v) for k, v in load_json("cities", "us-nws.json").items()
}
_DE_CITIES: dict[str, tuple[float, float]] = {
    k: tuple(v) for k, v in load_json("cities", "de-dwd.json").items()
}
_METOFFICE_CITIES: dict[str, tuple[float, float]] = {
    k: tuple(v) for k, v in load_json("cities", "metoffice.json").items()
}

_WMO_CODE_MAP: dict[int, WeatherCondition] = {
    int(k): WeatherCondition(v)
    for k, v in load_json("condition_maps", "wmo-codes.json").items()
}

_BASE_URL = "https://api.open-meteo.com/v1/forecast"
_CURRENT_PARAMS = (
    "temperature_2m,relative_humidity_2m,weather_code,wind_speed_10m,apparent_temperature"
)
_DAILY_PARAMS = (
    "weather_code,temperature_2m_max,temperature_2m_min,"
    "precipitation_probability_max,sunrise,sunset,uv_index_max"
)


class OpenMeteoProvider(WeatherProvider):
    """
    Open-Meteo weather provider.

    Coverage: Global (via lat/lon lookup)
    API Key: None required
    Priority: 11 (zero-config catch-all fallback below OpenWeatherMap)
    """

    priority = 11
    supports_forecast = True
    supports_air_quality = False
    requires_api_key = False

    @property
    def name(self) -> str:
        return "open-meteo"

    def supports_location(self, location: Location) -> bool:
        return self._resolve_coords(location) is not None

    async def get_current(self, location: Location) -> WeatherData:
        coords = self._resolve_coords(location)
        if coords is None:
            raise LocationNotSupportedError(
                f"Open-Meteo: cannot resolve coordinates for: {location.raw}"
            )
        lat, lon = coords
        params = {
            "latitude": f"{lat:.4f}",
            "longitude": f"{lon:.4f}",
            "current": _CURRENT_PARAMS,
            "daily": _DAILY_PARAMS,
            "forecast_days": "1",
            "timezone": "UTC",
        }
        data = await self._fetch(params)
        return self._parse_current(location, data)

    async def get_forecast(
        self, location: Location, days: int = 7
    ) -> list[WeatherData]:
        coords = self._resolve_coords(location)
        if coords is None:
            raise LocationNotSupportedError(
                f"Open-Meteo: cannot resolve coordinates for: {location.raw}"
            )
        lat, lon = coords
        params = {
            "latitude": f"{lat:.4f}",
            "longitude": f"{lon:.4f}",
            "daily": _DAILY_PARAMS,
            "forecast_days": str(min(days, 16)),
            "timezone": "UTC",
        }
        data = await self._fetch(params)
        return self._parse_forecast(location, data, days)

    # ── Private helpers ────────────────────────────────────────────────

    def _resolve_coords(
        self, location: Location
    ) -> Optional[tuple[float, float]]:
        if location.latitude is not None and location.longitude is not None:
            return (location.latitude, location.longitude)
        n = location.normalized
        return (
            _CN_CITIES.get(n)
            or _US_NWS_CITIES.get(n)
            or _DE_CITIES.get(n)
            or _METOFFICE_CITIES.get(n)
        )

    async def _fetch(self, params: dict) -> dict:
        url = f"{_BASE_URL}?{urllib.parse.urlencode(params)}"
        loop = asyncio.get_running_loop()

        def fetch():
            req = urllib.request.Request(url)
            req.add_header("Accept", "application/json")
            try:
                with urllib.request.urlopen(req, timeout=15) as resp:
                    return json.loads(resp.read().decode("utf-8"))
            except urllib.error.HTTPError as e:
                raise ProviderError(f"Open-Meteo API error: HTTP {e.code}")
            except Exception as e:
                raise ProviderError(f"Open-Meteo request failed: {e}")

        return await loop.run_in_executor(None, fetch)

    def _parse_current(self, location: Location, data: dict) -> WeatherData:
        current = data.get("current", {})
        daily = data.get("daily", {})

        wmo_code = current.get("weather_code", 0)
        condition = _WMO_CODE_MAP.get(wmo_code, WeatherCondition.UNKNOWN)

        max_temps = daily.get("temperature_2m_max", [])
        min_temps = daily.get("temperature_2m_min", [])
        precip_chances = daily.get("precipitation_probability_max", [])
        uv_indices = daily.get("uv_index_max", [])
        sunrises = daily.get("sunrise", [])
        sunsets = daily.get("sunset", [])
        observed_at = datetime.now(timezone.utc)
        raw_time = current.get("time")
        if isinstance(raw_time, str):
            try:
                observed_at = datetime.fromisoformat(raw_time).replace(
                    tzinfo=timezone.utc
                )
            except (ValueError, TypeError):
                pass

        return WeatherData(
            location=location.normalized.title(),
            latitude=data.get("latitude"),
            longitude=data.get("longitude"),
            temperature=current.get("temperature_2m", 0.0),
            feels_like=current.get("apparent_temperature"),
            humidity=current.get("relative_humidity_2m"),
            wind_speed=current.get("wind_speed_10m"),
            temp_high=max_temps[0] if max_temps else None,
            temp_low=min_temps[0] if min_temps else None,
            precipitation_chance=precip_chances[0] if precip_chances else None,
            uv_index=uv_indices[0] if uv_indices else None,
            sunrise=_format_time(sunrises[0]) if sunrises else None,
            sunset=_format_time(sunsets[0]) if sunsets else None,
            condition=condition,
            condition_raw=f"wmo:{wmo_code}",
            observed_at=observed_at,
            provider_name=self.name,
        )

    def _parse_forecast(
        self, location: Location, data: dict, days: int
    ) -> list[WeatherData]:
        daily = data.get("daily", {})
        times: list[str] = daily.get("time", [])
        wmo_codes: list[int] = daily.get("weather_code", [])
        max_temps: list[float] = daily.get("temperature_2m_max", [])
        min_temps: list[float] = daily.get("temperature_2m_min", [])
        precip_chances: list[int] = daily.get("precipitation_probability_max", [])
        uv_indices: list[float] = daily.get("uv_index_max", [])
        sunrises: list[str] = daily.get("sunrise", [])
        sunsets: list[str] = daily.get("sunset", [])

        location_name = location.normalized.title()
        forecasts = []

        for i, date_str in enumerate(times[:days]):
            wmo_code = wmo_codes[i] if i < len(wmo_codes) else 0
            condition = _WMO_CODE_MAP.get(wmo_code, WeatherCondition.UNKNOWN)
            forecasts.append(
                WeatherData(
                    location=location_name,
                    temperature=0.0,
                    condition=condition,
                    condition_raw=f"wmo:{wmo_code}",
                    forecast_date=date.fromisoformat(date_str),
                    temp_high=max_temps[i] if i < len(max_temps) else None,
                    temp_low=min_temps[i] if i < len(min_temps) else None,
                    precipitation_chance=(
                        precip_chances[i] if i < len(precip_chances) else None
                    ),
                    uv_index=uv_indices[i] if i < len(uv_indices) else None,
                    sunrise=_format_time(sunrises[i]) if i < len(sunrises) else None,
                    sunset=_format_time(sunsets[i]) if i < len(sunsets) else None,
                    provider_name=self.name,
                )
            )
        return forecasts


def _format_time(iso: str) -> str:
    """Extract HH:MM from an ISO 8601 timestamp string.

    >>> _format_time("2026-01-01T06:52")
    '06:52'
    """
    m = _TIME_RE.search(iso)
    return m.group(1) if m else iso


import re as _re
_TIME_RE = _re.compile(r"(\d{2}:\d{2})")
